NBXplorer/Repository.cs

```diff
@@ -1149,67 +1216,103 @@ public Task<TrackedTransaction[]> GetMatches(NBitcoin.Transaction tx, uint256 bl
 		{
 			foreach (var tx in txs)
 				tx.PrecomputeHash(false, true);
 			var transactionsPerScript = new MultiValueDictionary<Script, NBitcoin.Transaction>();
+			var transactionsPerOutpoint = new MultiValueDictionary<OutPoint, NBitcoin.Transaction>();
 			var matches = new Dictionary<string, TrackedTransaction>();
-			HashSet<Script> scripts = new HashSet<Script>(txs.Count);
 			var noMatchTransactions = new HashSet<uint256>(txs.Count);
+			var outPointsWithScripts = new HashSet<(OutPoint outPoint, Script script)>();
 			foreach (var tx in txs)
 			{
 				if (blockId != null && useCache && noMatchCache.Contains(tx.GetHash()))
 				{
 					continue;
 				}
 				noMatchTransactions.Add(tx.GetHash());
-				foreach (var input in tx.Inputs)
-				{
-					var signer = input.GetSigner();
-					if (signer != null)
-					{
-						scripts.Add(signer.ScriptPubKey);
-						transactionsPerScript.Add(signer.ScriptPubKey, tx);
-					}
-				}
-				foreach (var output in tx.Outputs)
+				var txOutPointsWithScripts = ExtractInfo(tx);
+
+				foreach (var outPointWithScript in txOutPointsWithScripts)
 				{
-					if (MinUtxoValue != null && output.Value < MinUtxoValue)
-						continue;
-					scripts.Add(output.ScriptPubKey);
-					transactionsPerScript.Add(output.ScriptPubKey, tx);
+					transactionsPerOutpoint.Add(outPointWithScript.outPoint, tx);
 				}
+				outPointsWithScripts.AddRange(txOutPointsWithScripts);
 			}
-			if (scripts.Count == 0)
-				return Array.Empty<TrackedTransaction>();
+
 			var keyPathInformationsByTrackedTransaction = new MultiValueDictionary<TrackedTransaction, KeyPathInformation>();
-			var keyInformations = await GetKeyInformations(scripts.ToArray());
-			foreach (var keyInfoByScripts in keyInformations)
+
+			if (outPointsWithScripts.Count > 0)
 			{
-				foreach (var tx in transactionsPerScript[keyInfoByScripts.Key])
+				var keyInformations = await GetKeyInformations(outPointsWithScripts.ToArray());
+				foreach (var keyInfoByOutpoints in keyInformations)
 				{
-					if (keyInfoByScripts.Value.Count != 0)
-						noMatchTransactions.Remove(tx.GetHash());
-					foreach (var keyInfo in keyInfoByScripts.Value)
-					{
-						var matchesGroupingKey = $"{keyInfo.DerivationStrategy?.ToString() ?? keyInfo.ScriptPubKey.ToHex()}-[{tx.GetHash()}]";
-						if (!matches.TryGetValue(matchesGroupingKey, out TrackedTransaction match))
-						{
-							match = CreateTrackedTransaction(keyInfo.TrackedSource,
-								new TrackedTransactionKey(tx.GetHash(), blockId, false),
-								tx,
-								new Dictionary<Script, KeyPath>());
-							match.FirstSeen = now;
-							match.Inserted = now;
-							matches.Add(matchesGroupingKey, match);
-						}
-						if (keyInfo.KeyPath != null)
-							match.KnownKeyPathMapping.TryAdd(keyInfo.ScriptPubKey, keyInfo.KeyPath);
-						keyPathInformationsByTrackedTransaction.Add(match, keyInfo);
-					}
+					recordMatches(
```

Comment:
`recordMatches` is not needed, we don't reuse it.

Reply:
Fidex this.